centreon/src/Core/HostTemplate/Domain/Model/MacroManagement.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+/*
+ * Copyright 2005 - 2023 Centreon (https://www.centreon.com/)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+declare(strict_types=1);
+
+namespace Core\HostTemplate\Domain\Model;
+
+use Core\CommandMacro\Domain\Model\CommandMacro;
+use Core\HostMacro\Domain\Model\HostMacro;
+
+// TODO : naming management/service/helper/... ?
+class MacroManagement
+{
+    /**
+     * @param HostMacro[] $macros
+     * @return array{
+     *      array<string,HostMacro>,
+     *      array<string,HostMacro>
+     * }
+     */
+    public static function resolveInheritanceForHostMacro(array $macros, array $inheritanceLine, int $childId): array
+    {
+        /** @var array<string,HostMacro> */
+        $inheritedMacros = [];
+        /** @var array<string,HostMacro> */
+        $directMacros = [];
+        foreach ($inheritanceLine as $parentId) {
+            foreach ($macros as $macro) {
+                if ($macro->getHostId() === $childId) {
+                    $directMacros[$macro->getName()] = $macro;
+                } else if (
+                    ! isset($inheritedMacros[$macro->getName()])
+                    && $macro->getHostId() === $parentId
+                ) {
+                    $inheritedMacros[$macro->getName()] = $macro;
+                }
+            }
+        }
```

Comment:
Split the flow because the `$directMacro` check is uselessly done for each `$inheritanceLine` element.
And it is easier to understand the algo by splitting the production of each array.
```suggestion
        /** @var array<string,HostMacro> $directMacros */
        $directMacros = [];
        foreach ($macros as $macro) {
            if ($macro->getHostId() === $childId) {
                $directMacros[$macro->getName()] = $macro;
            }
        }

        /** @var array<string,HostMacro> $inheritedMacros */
        $inheritedMacros = [];
        foreach ($inheritanceLine as $parentId) {
            foreach ($macros as $macro) {
                if (
                    ! isset($directMacros[$macro->getName()])
                    && ! isset($inheritedMacros[$macro->getName()])
                    && $macro->getHostId() === $parentId
                ) {
                    $inheritedMacros[$macro->getName()] = $macro;
                }
            }
        }
```